stream/stream.go

```diff
@@ -0,0 +1,137 @@
+package stream
+
+import (
+	"encoding/gob"
+	"errors"
+	"net/http"
+
+	"github.com/google/zoekt"
+	"github.com/google/zoekt/query"
+	"github.com/google/zoekt/rpc"
+)
+
+const DefaultSSEPath = "/stream"
+
+func Server(searcher zoekt.Searcher) http.Handler {
+	rpc.RegisterGob()
+	return &streamHandler{Searcher: searcher}
+}
+
+type searchArgs struct {
+	Q    query.Q
+	Opts *zoekt.SearchOptions
+}
+
+type searchReply struct {
+	Event  string
+	Result *zoekt.SearchResult
+}
+
+type streamHandler struct {
+	Searcher zoekt.Searcher
+}
+
+func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// Decode payload.
+	args := new(searchArgs)
+	err := gob.NewDecoder(r.Body).Decode(args)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	eventWriter, err := newEventStreamWriter(w)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Always send a done event in the end.
+	defer func() {
+		err = eventWriter.Event("done", nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	}()
+
+	// Kick-off search (in batch-mode for now).
+	searchResults, err := h.Searcher.Search(ctx, args.Q, args.Opts)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// We simulate streaming by sending searchResults in chunks over the wire. Later
+	// we want Searcher.Search to take a channel, buffer here and send chunks over
+	// the network.
+	var chunk *zoekt.SearchResult
+	chunk = &zoekt.SearchResult{
+		Stats:         searchResults.Stats,
+		Files:         nil,
+		RepoURLs:      searchResults.RepoURLs,
+		LineFragments: searchResults.LineFragments,
```

Comment:
we don't use these fields so we can leave them out.

Reply:
We aggregate them in `horizontal.go` but we don't read them ever. Will remove it here and then in Sourcegraph, too.